### PR TITLE
Refactor `VarNode`

### LINKED ISSUE #26820: WebGPU Node system extremely slow startup
### Description

Hi, I started exploring webgpu in threejs, however the startup time is just insane,
in the following scene with 4 materials, the scene only shows up after 1.6 seconds.
 
After a bit a profiling, it seems like the culprit is the WGSLNodeBuilder: each material has about 1000 nodes to construct which takes forever (about 400+ms per material). 

https://threejs.org/examples/?q=webgpu#webgpu_clearcoat

Any fixes on the table? Is this going to be a limitation of the node system

At the moment it is not really usable beyond small prototypes but even then having the wait multiple seconds for those node materials to build is definitely not ideal

### Reproduction steps

1. Open https://threejs.org/examples/?q=webgpu#webgpu_clearcoat
2. Profile in the dev console

### Screenshots

Screenshot 2023-09-21 165039

### Version

156

### Device

Desktop

### Browser

Chrome

### OS

Windows

### PR DESCRIPTION
**Pull request by LeviPesin**
Fixes https://github.com/mrdoob/three.js/issues/26820

Also helps https://github.com/mrdoob/three.js/pull/26795

**Description**
Refactor `VarNode` (also move some defaults from `VarNode.generate()` to `NodeBuilder.getVarFromNode()`). Also remove `replaceNode` argument from `Node.traverse()` callback signature.

~~Currently this (more specifically, lighting) **does not work** -- I think it is so because the nodes from lighting context are referencing nodes from the builder's stack which have not been built to that time yet. I tried to fix this by explicitly building the stack in ContextNode, but it still didn't work, and I don't really understand stacks and contexts theme deep enough to debug it... @sunag Can you please help with this?~~

### CODE REVIEW

**Comment by sunag** [comment]
We need a `ShaderCallNode` like happen in `FunctionCallNode`, it will need to store the call arguments and fire only in the `construct` process, in this way we can pass the `StackNode`.

**Comment by LeviPesin** [comment]
Can you please add it to this PR?

**Comment by sunag** [comment]
> Can you please add it to this PR?

I've already started developing it, as soon as it's ready I'll publish it, I'm doing it in a new PR to make it easier.

**Comment by LeviPesin** [comment]
@sunag Can you please tell how to fix this PR with the new change?

**Comment by sunag** [comment]
There are more changes to be made here, as soon as possible I will publish a new PR.

File: examples/jsm/nodes/lighting/HemisphereLightNode.js [hunk 1]
```diff
@@ -42,7 +42,7 @@ class HemisphereLightNode extends AnalyticLightNode {
 
 		const irradiance = mix( groundColorNode, colorNode, hemiDiffuseWeight );
 
-		builder.context.irradiance.addAssign( irradiance );
+		builder.stack.addAssign( builder.context.irradiance, irradiance );
```
Comment (LeviPesin) [inline examples/jsm/nodes/lighting/HemisphereLightNode.js:45]: @sunag I think you missed this in your PR?

**Comment by sunag** [comment]
Thanks!

**Pull request merged**